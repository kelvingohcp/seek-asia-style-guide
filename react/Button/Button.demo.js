import React from 'react';
import { Button, HeartIcon, StarIcon, Icon } from 'seek-asia-style-guide/react';
import * as sketch from './Button.sketch';
import styles from './Button.less';
import classnames from 'classnames';

export default {
  route: '/button',
  title: 'Button',
  component: Button,
  sketch,
  initialProps: {
    children: 'Hello world',
    color: 'callToAction'
  },
  options: [
    {
      label: 'States',
      type: 'checklist',
      states: [
        {
          label: 'Hover',
          transformProps: ({ className, ...props }) => ({
            ...props,
            className: classnames(className, styles.rootHover)
          })
        },
        {
          label: 'Active',
          transformProps: ({ className, ...props }) => ({
            ...props,
            className: classnames(className, styles.rootActive)
          })
        },
        {
          label: 'isSelected',
          transformProps: ({ className, ...props }) => ({
            ...props,
            isSelected: true
          })
        },
        {
          label: 'Compact',
          transformProps: props => ({
            ...props,
            compact: true
          })
        },
        {
          label: 'Disabled',
          transformProps: props => ({
            ...props,
            disabled: true
          })
        }
      ]
    },
    {
      label: 'Component',
      type: 'radio',
      states: [
        {
          label: 'As Button',
          transformProps: props => ({
            ...props,
            component: 'button'
          })
        },
        {
          label: 'As Link',
          transformProps: props => ({
            ...props,
            component: 'a',
            href: '#'
          })
        },
        {
          label: 'As Div',
          transformProps: props => ({
            ...props,
            component: 'div'
          })
        }
      ]
    },
    {
      label: 'Colour',
      type: 'radio',
      states: [
        {
          label: 'Call To Action if no tenant on env variable or prop',
          transformProps: props => ({
            ...props,
            color: 'callToAction'
          })
        },
        {
          label: 'Call To Action(JobStreet)',
          transformProps: props => ({
            ...props,
            color: 'callToAction',
            isJobStreet: true
          })
        },
        {
          label: 'Call To Action(JobsDB)',
          transformProps: props => ({
            ...props,
            color: 'callToAction',
            isJobsDB: true
          })
        },
        {
          label: 'Primary',
          transformProps: props => ({
            ...props,
            color: 'primary'
          })
        },
        {
          label: 'Secondary',
          transformProps: props => ({
            ...props,
            color: 'secondary'
          })
        },
        {
          label: 'Hyperlink',
          transformProps: props => ({
            ...props,
            color: 'hyperlink'
          })
        },
        {
          label: 'Tertiary',
          transformProps: props => ({
            ...props,
            color: 'tertiary'
          })
        },
        {
          label: 'Tertiary - DB Hirer',
          transformProps: props => ({
            ...props,
            color: 'tertiary',
            isJobsDB: true
          })
        },
        {
          label: 'Tertiary - JS Hirer (SiVA)',
          transformProps: props => ({
            ...props,
            color: 'tertiary',
            isJobStreet: true
          })
        }
      ]
    },
    {
      label: 'Icon',
      type: 'radio',
      states: [
        {
          label: 'No Icon',
          transformProps: props => props
        },
        {
          label: 'Heart Icon',
          transformProps: props => ({
            ...props,
            children: (
              <span>
                <HeartIcon filled={true} />
                &nbsp;Click here
              </span>
            )
          })
        },
        {
          label: 'Star Icon',
          transformProps: props => ({
            ...props,
            children: (
              <span>
                <StarIcon filled={true} />
                &nbsp;Click here
              </span>
            )
          })
        },
        {
          label: 'Medical',
          transformProps: props => ({
            ...props,
            children: (
              <span>
                <Icon type="medical" size="small" animation="popIn" lineHeight="waving" />
                &nbsp;Click here
              </span>
            )
          })
        },
        {
          label: 'Past Application',
          transformProps: props => ({
            ...props,
            children: (
              <span>
                <Icon type="pastApplication" size="small" animation="popIn" lineHeight="waving" />
                &nbsp;Click here
              </span>
            )
          })
        }
      ]
    }
  ]
};
