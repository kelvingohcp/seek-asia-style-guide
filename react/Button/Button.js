import styles from './Button.less';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import * as tenant from '../private/tenant';

export default class Button extends Component {
  static displayName = 'Button';

  static propTypes = {
    'aria-pressed': PropTypes.bool,
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node
    ]).isRequired,
    className: PropTypes.string,
    color: PropTypes.oneOf([
      'callToAction', 'primary', 'secondary', 'hyperlink', 'tertiary'
    ]).isRequired,
    compact: PropTypes.bool,
    component: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.string
    ]),
    disabled: PropTypes.bool,
    isJobsDB: PropTypes.bool,
    isJobStreet: PropTypes.bool,
    isSelected: PropTypes.bool,
    role: PropTypes.string,
    tabIndex: PropTypes.number
  };

  static defaultProps = {
    'aria-pressed': false,
    className: '',
    component: 'button',
    role: 'button',
    tabIndex: 0
  };

  storeButtonReference = button => {
    if (button !== null) {
      this.button = button;
    }
  }

  render() {
    const { color, compact, className, children, component, disabled, isJobsDB, isJobStreet, isSelected, ...restProps } = this.props;

    const combinedProps = {
      className: classnames(styles.root, className, {
        [styles.compact]: compact,
        [styles.disabled]: disabled,
        [styles.jobsDB]: isJobsDB || tenant.isJobsDB,
        [styles.jobStreet]: isJobStreet || tenant.isJobStreet,
        [styles.root_callToAction]: color === 'callToAction',
        [styles.root_hyperlink]: color === 'hyperlink',
        [styles.root_primary]: color === 'primary',
        [styles.root_secondary]: color === 'secondary',
        [styles.root_tertiary]: color === 'tertiary',
        [styles.selected]: isSelected
      }),
      ref: this.storeButtonReference,
      ...restProps
    };

    return React.createElement(component, combinedProps, children);
  }
}
