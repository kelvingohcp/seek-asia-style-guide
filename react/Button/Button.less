@import (reference) '~seek-asia-style-guide/theme';

.hoverState(@rules) {
  &:hover,
  &.rootHover {
    @rules();
  }
}

.activeState(@rules) {
  &:active,
  &.rootActive {
    @rules();
  }
}

.buttonColor(@color, @background-color) {
  color: @color;
  background-color: @background-color;
  .hoverState({background-color: darken(@background-color, 5%) ;});
}

.root {
  .touchableText(@waving-type-scale);
  display: inline-block;
  text-decoration: none;
  user-select: none;
  text-align: center;
  padding: 0 @grid-base * 4;
  border: 0;
  border-radius: 3px;
  transition:
    background-color 0.2s,
    transform 0.2s cubic-bezier(0.02, 1.505, 0.745, 1.235);
  cursor: pointer;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: bold;
  .activeState({outline: none;});
}

.compact {
  .compactTouchableText(@whistling-type-scale);
}

.disabled:not(.root_hyperlink) {
  .buttonColor(@saGrey5, @saGrey4);
  cursor: default;
}

.root_callToAction {
  .buttonColor(@saWhite, @saBlue);
}

.jobStreet.root_callToAction {
  .buttonColor(@sk-black, @actionYellow);
  &.disabled {
    .buttonColor(@saGrey5, @saGrey4);
    cursor: default;
  }
}

.jobsDB.root_callToAction {
  .buttonColor(@saWhite, @actionOrange);
  &.disabled {
    .buttonColor(@saGrey5, @saGrey4);
    cursor: default;
  }
}

.root_primary {
  .buttonColor(@saWhite, @actionBlue);
}

.root_secondary {
  .buttonColor(@actionBlue, @actionPastelBlue);
}

.root_hyperlink {
  .touchableText(@conversational-type-scale);
  color: @hyperlink;
  background-color: transparent;
  padding-left: 0;
  padding-right: 0;

  .hoverState({text-decoration: underline;});

  .activeState({transform: none;});

  &.disabled {
    color: @saGrey4;
    text-decoration: underline;
    cursor: default;
  }
}

.root_hyperlink.compact {
  .compactTouchableText(@whistling-type-scale);
}

.root_tertiary {
  .buttonColor(@actionTurquoise, @saWhite);

  .hoverState({background-color: @actionTurquoise; color: @saWhite;});
  border: 1px solid @actionTurquoise;
  padding-bottom: 10px;
}

.root_tertiary.compact {
  padding-bottom: 7px;
}

.group {
  width: 100%;

  .root {
    vertical-align: top;
    @media only screen and (max-width: 500px) {
      width: 100%;
    }
    &:not(.root_hyperlink) {
      @media @mobile {
        padding-left: @grid-gutter-width * 2;
        padding-right: @grid-gutter-width * 2;
      }
    }
    & + .root {
      margin-left: @grid-gutter-width / 2;
      @media only screen and (max-width: 500px) {
        margin-top: @grid-row-height;
        margin-left: 0;
      }
    }
    & + .root_hyperlink {
      margin-left: @grid-gutter-width;
      @media only screen and (max-width: 500px) {
        margin-left: 0;
      }
    }
  }
}
