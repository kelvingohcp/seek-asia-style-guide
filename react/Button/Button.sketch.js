import React from 'react';
import Button from './Button';

export const symbols = {
  'Button/Call To Action/Standard': (
    <Button color="callToAction">Button</Button>
  ),
  'Button/Call To Action/Compact': (
    <Button color="callToAction" compact>
      Button
    </Button>
  ),
  'Button/Call To Action/JobStreet/Standard': (
    <Button color="callToAction" isJobStreet>
      Button
    </Button>
  ),
  'Button/Call To Action/JobStreet/Compact': (
    <Button color="callToAction" isJobStreet compact>
      Button
    </Button>
  ),
  'Button/Call To Action/JobsDB/Standard': (
    <Button color="callToAction" isJobsDB>
      Button
    </Button>
  ),
  'Button/Call To Action/JobsDB/Compact': (
    <Button color="callToAction" isJobsDB compact>
      Button
    </Button>
  ),
  'Button/Primary/Standard': <Button color="primary">Button</Button>,
  'Button/Primary/Compact': (
    <Button color="primary" compact>
      Button
    </Button>
  ),
  'Button/Secondary/Standard': <Button color="secondary">Button</Button>,
  'Button/Secondary/Compact': (
    <Button color="secondary" compact>
      Button
    </Button>
  ),
  'Button/Hyperlink/Standard': <Button color="hyperlink">Button</Button>,
  'Button/Hyperlink/Compact': (
    <Button color="hyperlink" compact>
      Button
    </Button>
  ),
  'Button/Tertiary/jobsDB Hirer/Standard': (
    <Button color="tertiary">Button</Button>
  ),
  'Button/Tertiary/jobsDB Hirer/Compact': (
    <Button color="tertiary" compact>
      Button
    </Button>
  )
};
