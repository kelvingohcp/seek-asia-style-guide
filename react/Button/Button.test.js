import React from 'react';
import { shallow } from 'enzyme';
import Button from './Button';

describe('Button:', () => {
  describe('color variants:', () => {
    it('should render callToAction button', () => {
      const wrapper = shallow(<Button color="callToAction">SEEK</Button>);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render JobsDB callToAction button', () => {
      const wrapper = shallow(
        <Button isJobsDB color="callToAction">
          SEEK
        </Button>
      );
      expect(wrapper).toMatchSnapshot();
    });

    it('should render JobStreet callToAction button', () => {
      const wrapper = shallow(
        <Button isJobStreet color="callToAction">
          SEEK
        </Button>
      );
      expect(wrapper).toMatchSnapshot();
    });

    it('should render hyperlink color button', () => {
      const wrapper = shallow(<Button color="hyperlink">SEEK</Button>);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render primary button', () => {
      const wrapper = shallow(<Button color="primary">SEEK</Button>);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render secondary button', () => {
      const wrapper = shallow(<Button color="secondary">SEEK</Button>);
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('should render with node', () => {
    const wrapper = shallow(
      <Button color="callToAction">
        <h5>SEEK</h5>
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with array of nodes', () => {
    const wrapper = shallow(
      <Button color="callToAction">
        <span>SEEK</span>
        <span>AU</span>
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render as an anchor with href="https://www.seek.com.au"', () => {
    const wrapper = shallow(
      <Button color="hyperlink" component="a" href="https://www.seek.com.au">
        SEEK
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with className', () => {
    const wrapper = shallow(
      <Button color="callToAction" className="foo">
        SEEK
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render compact font size', () => {
    const wrapper = shallow(
      <Button color="callToAction" compact>
        SEEK
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render custom component based on itâ€™s reference', () => {
    const CustomComponent = props => <a {...props} someProp="kraken" />;
    const wrapper = shallow(
      <Button color="callToAction" component={CustomComponent}>
        SEEK
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render disabled', () => {
    const wrapper = shallow(
      <Button color="callToAction" disabled>
        SEEK
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render as div', () => {
    const wrapper = shallow(
      <Button component="div" color="callToAction">
        SEEK
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

});
