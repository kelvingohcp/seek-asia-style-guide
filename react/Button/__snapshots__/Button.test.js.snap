// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button: color variants: should render JobStreet callToAction button 1`] = `
<button
  aria-pressed={false}
  className="root jobStreet root_callToAction"
  role="button"
  tabindex={0}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render JobsDB callToAction button 1`] = `
<button
  aria-pressed={false}
  className="root jobsDB root_callToAction"
  role="button"
  tabindex={0}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render callToAction button 1`] = `
<button
  aria-pressed={false}
  className="root root_callToAction"
  role="button"
  tabindex={0}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render hyperlink color button 1`] = `
<button
  aria-pressed={false}
  className="root root_hyperlink"
  role="button"
  tabindex={0}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render primary button 1`] = `
<button
  aria-pressed={false}
  className="root root_primary"
  role="button"
  tabindex={0}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render secondary button 1`] = `
<button
  aria-pressed={false}
  className="root root_secondary"
  role="button"
  tabindex={0}
>
  SEEK
</button>
`;

exports[`Button: should render as an anchor with href="https://www.seek.com.au" 1`] = `
<a
  aria-pressed={false}
  className="root root_hyperlink"
  href="https://www.seek.com.au"
  role="button"
  tabindex={0}
>
  SEEK
</a>
`;

exports[`Button: should render as div 1`] = `
<div
  aria-pressed={false}
  className="root root_callToAction"
  role="button"
  tabindex={0}
>
  SEEK
</div>
`;

exports[`Button: should render compact font size 1`] = `
<button
  aria-pressed={false}
  className="root compact root_callToAction"
  role="button"
  tabindex={0}
>
  SEEK
</button>
`;

exports[`Button: should render custom component based on itâ€™s reference 1`] = `
<CustomComponent
  aria-pressed={false}
  className="root root_callToAction"
  role="button"
  tabindex={0}
>
  SEEK
</CustomComponent>
`;

exports[`Button: should render disabled 1`] = `
<button
  aria-pressed={false}
  className="root disabled root_callToAction"
  role="button"
  tabindex={0}
>
  SEEK
</button>
`;

exports[`Button: should render with array of nodes 1`] = `
<button
  aria-pressed={false}
  className="root root_callToAction"
  role="button"
  tabindex={0}
>
  <span>
    SEEK
  </span>
  <span>
    AU
  </span>
</button>
`;

exports[`Button: should render with className 1`] = `
<button
  aria-pressed={false}
  className="root foo root_callToAction"
  role="button"
  tabindex={0}
>
  SEEK
</button>
`;

exports[`Button: should render with isSelected style 1`] = `
<button
  aria-pressed={false}
  className="root root_callToAction selected"
  role="button"
  tabindex={0}
>
  SEEK
</button>
`;

exports[`Button: should render with node 1`] = `
<button
  aria-pressed={false}
  className="root root_callToAction"
  role="button"
  tabindex={0}
>
  <h5>
    SEEK
  </h5>
</button>
`;
